pecos/core/ann/quantizer_impl/common.hpp

```diff
@@ -0,0 +1,237 @@
+            for (index_type d = 0; d < num_local_codebooks; d++) {
+                for (size_t k = 0; k < num_of_local_centroids; k++) {
+                    lut[d * num_of_local_centroids + k] = std::round((raw_dist[d * num_of_local_centroids + k] - bias) / scale);
+                }
+            }
+        }
+
+        inline void encode(float* query, uint8_t* codes) {
+            for (index_type d = 0; d < num_local_codebooks; d++) {
+                std::vector<float > results;
+                for (size_t k = 0; k < num_of_local_centroids; k++) {
+                    float v = 0;
+                    for (int j = 0; j < local_dimension; j++) {
+                        float tmp_v = original_local_codebooks[d * num_of_local_centroids * local_dimension + k * local_dimension + j]
+                            - (query[d * local_dimension + j] - global_centroid[d * local_dimension + j]);
+                        v += (tmp_v * tmp_v);
+                    }
+                    results.push_back(v);
+                }
+                std::vector<float>::iterator argmin_result = std::min_element(results.begin(), results.end());
+                codes[d] = std::distance(results.begin(), argmin_result);
+            }
+        }
+
+        inline void compute_centroids(pecos::drm_t& X, int dsub, size_t ksub, index_type *assign, float *centroids, int threads=1) {
+            // zero initialization for later do_axpy
+            memset(centroids, 0, ksub * dsub * sizeof(*centroids));
+            std::vector<float> centroids_size(ksub);
+            #pragma omp parallel num_threads(threads)
+            {
+                // each thread takes care of [c_l, c_r)
+                int rank = omp_get_thread_num();
+                size_t c_l = (ksub * rank) / threads;
+                size_t c_r = (ksub * (rank + 1)) / threads;
+                for (size_t i = 0; i < X.rows; i++) {
+                    auto ci = assign[i];
+                    if (ci >= c_l && ci < c_r) {
+                        float* y = centroids + ci * dsub;
+                        const auto& xi = X.get_row(i);
+                        pecos::do_axpy(1.0, xi.val, y, dsub);
+                        centroids_size[ci] += 1;
+                    }
+                }
+                // normalize center vector
+                for (size_t ci = c_l; ci < c_r; ci++) {
+                    float* y = centroids + ci * dsub;
+                    pecos::do_scale(1.0 / centroids_size[ci], y, dsub);
+                }
+            }
+        }
+
+        inline void train(const pecos::drm_t& X_trn, int M, size_t sub_sample_points=0, int seed=0, size_t max_iter=10, int threads=32) {
+            size_t dimension = X_trn.cols;
+            if (dimension % M != 0) {
+                throw std::runtime_error("Original dimension must be divided by subspace dimension");
+            }
+            num_local_codebooks = M;
```

Comment:
consider just replace `M` by `num_local_codebooks`, and use that naming consistently in other place.

Reply:
Resolve in the next commit.